api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java

```diff
@@ -87,7 +87,148 @@ public void onShutdown() {
 	 * @see org.openmrs.api.ObsService#saveObs(org.openmrs.Obs, String)
 	 */
 	public Obs saveObs(Obs obs, String changeMessage) throws APIException {
-		if (null != obs && null != obs.getConcept() && obs.getConcept().isComplex()
+		if(obs == null){
+			throw new APIException("obs cannot be null!!", (Object[]) null); //TODO: change this to a message string!!
+		}
+
+		if(obs.getId() != null && changeMessage == null){
+			throw new APIException("Obs.error.ChangeMessage.required", (Object[]) null);
+		}
+
+		handleExistingObsWithComplexConcept(obs);
+
+		ensureRequirePrivilege(obs);
+
+		//Should allow updating a voided Obs, it seems to be pointless to restrict it,
+		//otherwise operations like merge patients won't be possible when to moving voided obs
+		if (obs.getObsId() == null || obs.getVoided()) {
+			return saveNewAndVoidedObs(obs,changeMessage);
+		} else if(!obs.isDirty()){
+			setPersonFromEncounter(obs);
```

Comment:
Agreed with you!  But this PR is as-is behavior of the [current code](https://github.com/openmrs/openmrs-core/blob/2.0.x/api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java#L127) and I just did the refactored it.  The next PR will be on the basis of this.  I will clean it up there.   We are setting up the Person on the obs irrespective of if the obs is dirty.